### PR TITLE
Elevated rate limits incremental quota

### PR DESCRIPTION
**Pull request by pubalokta**
### Description

This PR changes the way we track ERL quota to address an edge case where the ERL quota changes in the middle of the month (before quota reset at the beginning of the next month).

It also collects all ERL configuration in a single place, so that the bucket overrides functionality can also be applied to ERL configuration.

**Before**:
* The first time ERL was activated, it was setting in redis quotaKey = quotaGivenThroughParam . If quota=192, then quotaKey = 192 after that first activation.
* From that point on, every ERL activation it was basically decreasing the quota quotaKey = quotaKey-1
Until quotaKey=0 => then the tenant exhausted all the quota for the given month.
* With the current situation, if the quota configuration changes in the middle of the month, lets say from 192 to 384, that value will not be used anymore in the code until the quoteKey expires, which is at the end of month.

**After the changes in this PR:**
* The first time ERL is activated, it sets quotaKey = 1 .
* From that point on, every ERL activation it increases the quota: quotakey = quotaKey+1 until quotaKey=quotaGivenThroughParam. At that point the tenant exhausted the quota for the given month.
If the tenant gets its quota increased, then quotaGivenThroughParam will be higher, and then quotaKey=quotaGivenThroughParam  will be false, hence ERL will be activate.

### Testing

- [x] This change adds test coverage for new/changed/fixed functionality

### Checklist

- [x] I have added documentation for new/changed functionality in this PR or in auth0.com/docs
- [x] All active GitHub checks for tests, formatting, and security are passing
- [x] The correct base branch is being used, if not the default branch

### CODE REVIEW

File: lib/take_elevated.lua [hunk 1]
```diff
@@ -48,14 +47,15 @@ local function calculateNewBucketContent(current, tokens_per_ms, bucket_size, cu
 end
 
 local function takeERLQuota(erl_quota_key, erl_quota_amount, erl_quota_expiration_epoch)
```
Comment (LeweyM) [inline lib/take_elevated.lua:49]: I find mutable variables hard to follow, could we change this to have explicit returns for each case? Something like the following;

```suggestion
local function takeERLQuota(erl_quota_key, erl_quota_amount, erl_quota_expiration_epoch)
    if erl_quota_amount <= 0 then
        return 0
    end

    local get_quota_result = redis.call('GET', erl_quota_key)
    if type(get_quota_result) ~= 'string' then
        redis.call('SET', erl_quota_key, 1, 'PXAT', string.format('%.0f', erl_quota_expiration_epoch))
        return 1
    end

    local erl_quota = tonumber(get_quota_result)

    if erl_quota >= erl_quota_amount then
        return erl_quota
    end

    local incremented_erl_quota = erl_quota + 1
    redis.call('SET', erl_quota_key, incremented_erl_quota, 'PXAT', string.format('%.0f', erl_quota_expiration_epoch))
    return incremented_erl_quota
end
```

File: lib/take_elevated.lua [hunk 2]
```diff
@@ -48,14 +47,15 @@ local function calculateNewBucketContent(current, tokens_per_ms, bucket_size, cu
...
     end
 
-    if erl_quota > 0 then
-        redis.call('SET', erl_quota_key, erl_quota -1, 'PXAT', string.format('%.0f', erl_quota_expiration_epoch))
+    if erl_quota < erl_quota_amount then
+        erl_quota = erl_quota + 1
+        redis.call('SET', erl_quota_key, erl_quota, 'PXAT', string.format('%.0f', erl_quota_expiration_epoch))
     end
     return erl_quota
 end
```
Comment (LeweyM) [inline lib/take_elevated.lua:61]: I find mutable variables hard to follow, could we change this to have explicit returns for each case? Something like the following;

```suggestion
local function takeERLQuota(erl_quota_key, erl_quota_amount, erl_quota_expiration_epoch)
    if erl_quota_amount <= 0 then
        return 0
    end

    local get_quota_result = redis.call('GET', erl_quota_key)
    if type(get_quota_result) ~= 'string' then
        redis.call('SET', erl_quota_key, 1, 'PXAT', string.format('%.0f', erl_quota_expiration_epoch))
        return 1
    end

    local erl_quota = tonumber(get_quota_result)

    if erl_quota >= erl_quota_amount then
        return erl_quota
    end

    local incremented_erl_quota = erl_quota + 1
    redis.call('SET', erl_quota_key, incremented_erl_quota, 'PXAT', string.format('%.0f', erl_quota_expiration_epoch))
    return incremented_erl_quota
end
```

File: lib/validation.js [hunk 3]
```diff
@@ -1,5 +1,13 @@
 const { INTERVAL_SHORTCUTS, ERL_QUOTA_INTERVALS_SHORTCUTS } = require('./utils');
 
+const ERL_BUCKET_KEYS = [
+  'size', 'per_interval', 'erl_activation_period_seconds', 'erl_quota_amount', 'erl_quota_interval'
+];
+
+const ERL_MANDATORY_KEYS = [
+  'size', 'per_interval', 'erl_activation_period_seconds', 'quota_per_calendar_month'
```
[the remark this answers is not in the document]
Comment (LeweyM) [reply lib/validation.js:8]: I was confused by this too. Looking into it, it's because `ERL_MANDATORY_KEYS` is for the bucket config **before** parsing and `ERL_BUCKET_KEYS` is for the bucket config **after** parsing. That's also why `ERL_BUCKET_KEYS` does not fully match the docs, because it's our internal representation.

It's a bit confusing. I think we could simplify by validating the user input only. Wdyt?

File: test/utils.tests.js [hunk 4]
```diff
@@ -1,42 +1,152 @@
...
+        per_interval: 100,
+        ttl: 1,
+        ms_per_interval: 0.1,
+        erl_activation_period_seconds: 900,
+        erl_quota_amount: 0,
+        erl_quota_interval: 'quota_per_calendar_month',
+      });
+    });
+
+    it('should add overrides', () => {
```
[the remark this answers is not in the document]
Comment (pubalokta) [reply test/utils.tests.js:73]: there's a test for that in db.tests `should use elevated_limits config override when provided`.
But anyways I'm adding that specific test here

File: lib/validation.js [hunk 5]
```diff
@@ -1,5 +1,13 @@
 const { INTERVAL_SHORTCUTS, ERL_QUOTA_INTERVALS_SHORTCUTS } = require('./utils');
 
+const ERL_BUCKET_KEYS = [
+  'size', 'per_interval', 'erl_activation_period_seconds', 'erl_quota_amount', 'erl_quota_interval'
```
[the remark this answers is not in the document]
Comment (pubalokta) [reply lib/validation.js:4]: `quota_per_calendar_month` gets translated internally into:
- quota_amount
- quota_interval

File: test/utils.tests.js [hunk 6]
```diff
@@ -1,42 +1,152 @@
...
+      const bucket = {
+        size: 100,
+        per_second: 100,
+        overrides: {
```
[the remark this answers is not in the document]
Comment (pubalokta) [reply test/utils.tests.js:77]: added to README

File: lib/utils.js [hunk 7]
```diff
@@ -59,16 +64,36 @@ function normalizeElevatedTemporals(params) {
...
+
   return type;
 }
 
+function getElevatedParamsOrDefault(params) {
+  if (params.elevated_limits) {
+    return params.elevated_limits;
+  }
+
+  // as a default for elevated limits we'll use the same size and interval as the bucket
```
[the remark this answers is not in the document]
Comment (pubalokta) [reply lib/utils.js:88]: mentioned

File: lib/take_elevated.lua (hunk 2 above)
Comment (pubalokta) [reply lib/take_elevated.lua:61]: Done

**Review by ecita — APPROVED** [review]
